### PR TITLE
ADNI Classification using GFNet - s4763354

### PR DESCRIPTION
**Pull request by JoeyCheungly**
Hi Teaching Team,

This pull request is my implementation of ADNI classification project in the topic-recognition branch. This project aims to perform binary classification on brain images from the ADNI dataset into two categories: normal (NC) and Alzheimer's disease (AD) using GFNet. Please review the READ.md file for the model performance and visualizations, dependencies for training/testing and more. The files include: 

- modules.py: This file contains the code for creating the GFNet model network and its components.
- dataset.py:  This file contains data loader and preprocessing like train/val/test splits, patient ID splits, data augmnetation etc. Some transformed images example are also generated. 
- train.py: This file contains code for training, validating, testing, and saving the model. Plotting of training and validation loss and accuracy are present. 
- predict.py: This file is used to run inference/prediction of the model on two classes. It contains prediction on example inputs and confusion matrix.
- README.md: This is the detail project documentation.
- images/: This is a subdirectory for images and plots including visualizations, model architecture, predictions. 

Thank you for taking time to review my project. 

Lok Yee Joey Cheung 
s4763354

### CODE REVIEW

**Comment by wangzhaomxy** [comment]
**<This is an initial inspection, no action is required at this point.>**

**File Organizing:** Well-organized files.

**Problem Solving:**

- The algorithm solves the problem appropriately. 
- Accuracy in testing dataset: 0.75. 

**Model and functions:**

- It correctly uses PyTorch to construct the GFNet models and functions.
- With data augmentation.

**Code design:** Good.

**Code comment and docstring:**

- Good code comments 
- Minimal docstrings. It would be better with more function docstrings.

**Difficulty:** Hard.

**Additional Comments:**

- Tried and well discussed two ways of training the model: from scratch and based on the pre-trained model, which is a good attempt.
- Sufficient commits
- Good ReadMe design

**Comment by aniketgupta17** [comment]
Observational Feedback

Pull Request:
Correctly created the Pull request from Topic Recognition Branch .
The pull request includes a clear description about the file structure . 

File Organizing: Well-organized files.

Commit Log:
Commit messages could be more meaningful .
Commits demonstrate a logical development flow.
Commits are regularly made, showing incremental development rather than bulk updates.

Documentation:
The README file provides a clear explanation of the algorithm and the problem it solves.
Code comments and docstrings  are included and could be more informative .
Proper GitHub markdown formatting is used, with organized headings, lists, and code blocks.
Can add about future improvements in the ReadMe.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. Minimal doc strings.</td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. Uses Data Aug</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  GFNet (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-1</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
